lib/src/scene.dart

```diff
@@ -132,4 +132,9 @@ abstract class Scene {
   /// **and call [rebuildScene()] afterwards.**
   List<EnvironmentControlBuilder> environmentControlBuilders =
       <EnvironmentControlBuilder>[];
+
+  /// Called when [EnvironmentControlPanel]'s "reset" button is pressed.
+  /// Override this function to reset any state controlled by the Scene's
+  /// [environmentControlBuilders].
+  void onEnvironmentReset() {}
```

Comment:
I'm a bit worried about the scalability of the EnvironmentControl API now that it's leaking beyond EnvironmentControlBuilder. E.g. if we want to end up in a situation where there are a heap of prebuilt environment controls that people can easily install in different scenes, it might start to get finicky when you need to individually handle storing the state with ivars, resetting the state, maybe serialization etc later. Though, I do love that the current API is very simple for simple use cases and pretty low on boilerplate and I don't know if I have a better idea.

Alternative sketch. I dunno if I've reinvented a stateful widget or something here though... I thiiiiink the behavior should be identical to what you have now, just the management logic for each environment control is pushed into a child EnvironmentControl object rather than all mingling at the Scene level.

```
abstract class EnvironmentControl<T> {

Widget build(
  BuildContext context,
  VoidCallback rebuildScene,
);

void resetState();

SomeFormat serialize();

void deserialize(SomeFormat input)

T get value;

}

class MyScene {

// Each control has an object which individually stores its state.
final postNameControl = StringInputControl(name: "Post name", default: "my new post");

// The framework can iterate over the controls to do reset, serialization etc.
@override
List<EnvironmentControl> get environmentControls {
  return [ postNameControl ];
}

void build() {
  return Text (postNameControl.value)
}


}
```

Reply:
I know what you mean – I was starting to feel a bit iffy about this myself. This is an interesting proposal and not wildly different than an approach I explored when designing the custom controls API. I think this sidesteps some of the issues I ran into with Dart generics (namely, putting `MyClass<int>` and `MyClass<String>` into a list gives that list a type `[MyClass<dynamic>]` and you end up having to do a bunch of if/else branching to handle the individual cases).

I'll need to think about this some more in the context of state preservation/restoration. It will probably make sense for me to tackle that sooner rather than later.

Reply:
I still need to write tests, but how do you feel about landing this roughly as-is and overhauling the way environment state is managed (making it serializable, observable, etc. etc.) in a future PR that addresses https://github.com/bryanoltman/stager/issues/32?

Reply:
Agreed - let's just submit and refactor if/when necessary!